README.md
# terminal_layout

The project help you to quickly build layouts in terminal  
(这个一个命令行ui布局工具)

![demo_v2_1.gif](pic/demo_v2_1.gif)

<img src="pic/demo.gif"  alt="demo.gif" width="450"/>

----------------

**Some extensions base on terminal_layout**

* [progress](terminal_layout/extensions/progress/README.md)

![progress.gif](pic/progress.gif)

* [choice](terminal_layout/extensions/choice/README.md)

![choice.gif](pic/choice.gif)

-------------------

** video demo **

<a href="https://asciinema.org/a/226120">
<img src="https://asciinema.org/a/226120.svg"  alt="asciicast" width="550"/>
</a>


# link

* [Github](https://github.com/gojuukaze/terminal_layout) 
* [中文README](README.ZH.md) 
* [Documentation 中文文档](https://doc.ikaze.cn/terminal_layout/) 


# install 
```bash
pip install terminal-layout
```

# Dependencies
* Python 2.7, 3.5+ (maybe 3.4)
* Linux, OS X, and Windows systems.

# Usage

 * easy demo:

```python
import time
from terminal_layout import *

ctl = LayoutCtl.quick(TableLayout,
                      # table id: root
                      [
                          [TextView('t1', 'Hello World!', width=Width.fill, back=Back.blue)],  # <- row id: root_row_0,
                          [TextView('t2', '', fore=Fore.magenta)],  # <- row id: root_row_1,
                      ],
                      )

# or layout=ctl.get_layout()
layout = ctl.find_view_by_id('root')
layout.set_width(20)

# default: auto_re_draw=True
ctl.draw()

# call delay_set_text() must be set auto_re_draw=True,
# otherwise you must start a thread to call re_draw() by yourself
ctl.find_view_by_id('t2').delay_set_text('你好,世界!', delay=0.2)

time.sleep(0.5)
row3 = TableRow.quick_init('', [TextView('t3', 'こんにちは、世界!')])
layout.add_view(row3)

# If you call draw() with auto_re_draw=True, you must stop()
ctl.stop()

```
![](pic/hello.png)

* disable auto_re_draw

```python
import time
from terminal_layout import *

ctl = LayoutCtl.quick(TableLayout,
                      # table id: root
                      [
                          [TextView('t1', 'Hello World!', width=Width.fill, back=Back.blue)],  # <- row id: root_row_1,
                          [TextView('t2', '', fore=Fore.magenta)],  # <- row id: root_row_2,
                      ],
                      )


layout = ctl.find_view_by_id('root')
layout.set_width(20)

ctl.draw(auto_re_draw=False)

ctl.find_view_by_id('t2').set_text('你好,世界!')
ctl.re_draw()

time.sleep(0.5)
row3 = TableRow.quick_init('', [TextView('t3', 'こんにちは、世界!')])
layout.add_view(row3)
ctl.re_draw()

# don't need call stop()
# ctl.stop()
```

 * use python2 unicode

```python
# -*- coding: utf-8 -*-
from terminal_layout import *
import sys
reload(sys)
sys.setdefaultencoding('utf-8')

ctl = LayoutCtl.quick(TableLayout,
                      [
                          [TextView('', u'中文，你好', back=Back.cyan, width=Width.wrap)],
                          [TextView('', u'中文，你好', back=Back.cyan, width=6)],
                          [TextView('', u'日本語，こんにちは', back=Back.cyan, width=Width.wrap)],
                      ]

                      )

ctl.draw()

```

![](pic/py2.png)


## Properties

 * fore & back
 
```python
TextView('','fore',fore=Fore.red)
TextView('','back',back=Back.red)
```

<img width="560" src="pic/color.jpeg"/>

 * style

```python
TextView('','style',style=Style.dim)
```

<img width="560" src="pic/style.jpeg"/>

 * width
 
```python
TextView('','width',width=10)
```

<img width="560" src="pic/width.jpeg"/>

 * weight
 
```python
TextView('','weight',weight=1)
```

<img width="560" src="pic/weight.jpeg"/>

 * gravity
 
```python
TextView('','gravity',gravity=Gravity.left)
```

<img width="560" src="pic/gravity.jpeg"/>

 * visibility
 
```python
TextView('','',visibility=Visibility.visible)
```

<img width="560" src="pic/visibility.jpeg"/>


 * ex_style (not support windows)


```python
TextView('','ex_style',style=Style.ex_blink)
```

<img width="560" src="pic/ex_style.jpeg"/>

 * ex_fore & ex_back (not support windows)


```python
TextView('','ex_fore',fore=Fore.ex_red_1)
TextView('','ex_back',back=Back.ex_red_1)

```

<img width="560" src="pic/ex_color.jpeg"/>

# LICENSE

[GPLv3](https://github.com/gojuukaze/terminal_layout/blob/master/LICENSE)

# Thanks

 * [colorama](https://github.com/tartley/colorama) : Simple cross-platform colored terminal text in Python
 * [colored](https://gitlab.com/dslackw/colored) : Very simple Python library for color and formatting in terminal


# 捐赠 / Sponsor

开源不易，如果你觉得对你有帮助，求打赏个一块两块的

![](https://gitee.com/gojuukaze/liteAuth/raw/master/shang.jpg)



docs/changelog.rst
changelog
=============

2.1.4
---------
* ``TableLayout`` 添加 ``overflow_vertical`` 参数，用于terminal高度不够时隐藏row（默认不隐藏）
* 运行环境检测，非Terminal下抛出错误 ( `#25 <https://github.com/gojuukaze/terminal_layout/issues/25>`__ )
* 添加 `scroll <https://github.com/gojuukaze/terminal_layout/tree/master/terminal_layout/extensions/scroll>`__ 扩展，让 ``TableLayout`` 支持滚动 ( `#24 <https://github.com/gojuukaze/terminal_layout/issues/24>`__ )
* 添加 ``remove``, ``remove_view_by_id`` 函数
* choice扩展改用scroll实现滚动
* 添加 ``is_show`` 用于 使用 ``scroll``  或 ``overflow_vertical`` 为 ``hidden_top`` 、 ``hidden_btm`` 时判断 ``TableRow`` 是否隐藏。（  **只能判断TableRow** ）
* 修改一些小bug

2.1.3
---------
* 解决 ``readkey()`` 函数在 Win PowerShell 下无法识别方向键 bug ( `#22 <https://github.com/gojuukaze/terminal_layout/issues/22>`__ )
* choice扩展适配高度不够情况，当高度不够时隐藏部分选项 ( `#21 <https://github.com/gojuukaze/terminal_layout/issues/21>`__ )
* choice扩展支持设置stop_key，默认为 ``['q']``

2.1.2
---------
* 增加input扩展，可以获取文字输入了（不支持windows）
* TextView 增加 ``overflow`` 属性，用户文本过长时隐藏左边还是右边
* view 增加 parent 属性
* ctl自动重绘可通过设置 ``refresh_thread_stop`` 停止重绘


2.0.0
---------
* auto refresh 增加自动刷新功能
* add ``delay_set_text()`` 增加渐进显示字符的函数 ``delay_set_text()``
* ``find_view_by_id()`` 返回 ``ViewProxy`` ，不再直接返回view
* 增加扩展 extensions
* 增加按钮监听功能
* 修复python2 bug

1.0.0
--------


demo/record_demo.py
import time

from terminal_layout import *

ctl = LayoutCtl.quick(TableLayout,
                      [
                          [TextView('title', '** Terminal Layout **', fore=Fore.black, width=Width.fill,
                                    gravity=Gravity.center, style=Style.ex_bold, )],

                          [TextView('', 'What can terminal_layout do for you?')],
                          [TextView('', '')],
                          # 1. set fore back style in terminal
                          [TextView('s1-1', ''),
                           TextView('s1-2', '', fore=Fore.ex_dark_slate_gray_3),
                           TextView('s1-3', '', back=Back.ex_dark_slate_gray_3),
                           TextView('', ' '),
                           TextView('s1-4', '', style=Style.ex_underlined),
                           TextView('', ' '),
                           TextView('s1-5', ''),
                           ],
                          # 2. print string char one by one
                          [TextView('s2', '')],
                          # 3. listen keyboard events
                          [TextView('s3', '')],
                          # 4. draw a table
                          [TextView('s4', '')],
                          # 4-1. add row
                          [TextView('s4-1', '')],
                          # 4-2. change text
                          [TextView('s4-2', '')],
                          # 4-3. set gravity and more
                          [TextView('s4-3', '')],
                          [TextView('', '')],
                          # s3 table
                          [TextView('s4-table', 'Table', width=20, gravity=Gravity.center, back=Back.ex_slate_blue_3b,
                                    visibility=Visibility.invisible)],
                          [TextView('s4-table-r1-1', 'No1', width=10, back=Back.ex_sky_blue_2,
                                    visibility=Visibility.invisible),
                           TextView('s4-table-r1-2', 'Bob', width=10, back=Back.ex_sky_blue_2,
                                    visibility=Visibility.invisible)],
                          [TextView('s4-table-r2-1', 'No2', width=10, back=Back.ex_sky_blue_2,
                                    visibility=Visibility.invisible),
                           TextView('s4-table-r2-2', 'Tom', width=10, back=Back.ex_sky_blue_2,
                                    visibility=Visibility.invisible)],

                      ]

                      )


def s1():
    """
    1. set fore back style in terminal
    :return:
    :rtype:
    """
    for i, s in enumerate(["1. Set", " Fore ", " Back ", "Style", " in terminal"]):
        ctl.find_view_by_id('s1-%d' % (i + 1)).delay_set_text(s, delay=0.1)


def s2():
    """
    2. print string char one by one
    :return: 
    :rtype: 
    """
    ctl.find_view_by_id('s2').delay_set_text('2. Print string char one by one', delay=0.1)


def s3():
    """
    3. listen keyboard events
    :return: 
    :rtype: 
    """
    ctl.find_view_by_id('s3').delay_set_text('3. Listen keyboard events', delay=0.1)


def s4():
    """
    4. draw a table
    :return:
    :rtype:
    """

    ctl.find_view_by_id('s4').delay_set_text('4. Draw a table', delay=0.1)

    table_ids = ['s4-table', 's4-table-r1-1', 's4-table-r1-2', 's4-table-r2-1', 's4-table-r2-2']
    for v_id in table_ids:
        ctl.find_view_by_id(v_id).set_visibility(Visibility.visible)
    time.sleep(0.2)
    # 4-1. add row
    ctl.find_view_by_id('s4-1').delay_set_text('   4-1. add row', delay=0.1)
    ctl.get_layout().add_view(
        TableRow.quick_init('', [TextView('s4-table-r4-1', 'No3', width=10, back=Back.ex_sky_blue_2, ),
                                 TextView('s4-table-r4-2', '小王', width=10, back=Back.ex_sky_blue_2, )]))

    time.sleep(0.2)
    # 4-2. change text
    ctl.find_view_by_id('s4-2').delay_set_text('   4-2. change text', delay=0.1)
    ctl.find_view_by_id('s4-table-r4-2').set_text('Wong')
    ctl.find_view_by_id('s4-table-r4-2').set_fore(Fore.black)

    time.sleep(0.2)

    # 4-3. set gravity and more
    ctl.find_view_by_id('s4-3').delay_set_text('   4-3. set gravity and more', delay=0.1)

    for v_id in ['s4-table-r1-1', 's4-table-r1-2', 's4-table-r2-1', 's4-table-r2-2', 's4-table-r4-1', 's4-table-r4-2']:
        ctl.find_view_by_id(v_id).set_gravity(Gravity.center)


print('')
ctl.find_view_by_id('root_row_0').set_width(40)
ctl.draw()

time.sleep(0.5)
s1()
time.sleep(0.2)
s2()
time.sleep(0.2)
s3()
time.sleep(0.2)
s4()

ctl.stop()
print('')
